cluster-autoscaler/processors/emptycluster/empty_cluster_processor.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package emptycluster
+
+import (
+	"time"
+
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/autoscaler/cluster-autoscaler/clusterstate/utils"
+	"k8s.io/autoscaler/cluster-autoscaler/context"
+	"k8s.io/autoscaler/cluster-autoscaler/metrics"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	"k8s.io/klog/v2"
+)
+
+//NodesNotReadyAfterStartTimeout How long should Cluster Autoscaler wait for nodes to become ready after start.
+const NodesNotReadyAfterStartTimeout = 10 * time.Minute
+
+// EmptyClusterProcessor defines interface for whether action needs to be taken for the a cluster or not
+type EmptyClusterProcessor interface {
```

Comment:
I just realized the processor is always called and the interface is in no way specific to empty clusters. The default implementation happens to cover handling an empty cluster, but it could really check any preconditions for running a loop and abort if needed.

The issue here is that there are multiple such checks at the start of RunOnce (ex. https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/core/static_autoscaler.go#L344) and they happen at various places in RunOnce. This is due to historical reasons - other checks require clusterstate and the reason why CA needed special handling for empty clusters in the first place was that recalculating clusterstate couldn't handle those.
We could probably refactor the flow to handle them in one place now that clusterstate is fine with empty clusters, but that's a large refactor and we'd need to figure out who should own writing metrics, etc to really make it clean. That feels like more work than I should ask of you just to add this. But maybe there is some way to have more generic interface without large refactors? Let me think about it and I'll try to comment tomorrow morning.

Reply:
Just an idea:

```
type EarlyReturnProcessor interface {
    ShouldAbort(...) (bool, error)
    CleanUp()
}
```

Ideally, ShouldAbort() would only accept current context, since fetching node lists is possible from there, but since there is more going on there than just getting data from listers, it's probably safer to keep node lists in the arguments.